packages/dds/tree/src/domains/json/jsonCursor.ts

```diff
@@ -149,55 +49,42 @@ export class JsonCursor<T> implements ITreeCursor<SynchronousNavigationResult> {
                     return jsonObject.name;
                 }
         }
-    }
-
-    public get keys(): Iterable<FieldKey> {
-        const node = this.currentNode;
-        const type = typeof node;
-
-        switch (type) {
+    },
+    keysFromNode: (node: JsonCompatible): readonly FieldKey[] => {
+        switch (typeof node) {
             case "object":
                 if (node === null) {
                     return [];
                 } else if (Array.isArray(node)) {
                     return [EmptyKey];
                 } else {
-                    return Object.keys(node as object) as Iterable<FieldKey>;
+                    return Object.getOwnPropertyNames(node as object) as FieldKey[];
                 }
             default:
                 return [];
         }
-    }
-
-    public childFieldLength(key: FieldKey): number {
-        const node = this.currentNode;
-
-        // The 'Empty' field is used to access the indexer of array nodes.
+    },
+    getFieldFromNode: (node: JsonCompatible, key: FieldKey): readonly JsonCompatible[] => {
         if (key === EmptyKey && Array.isArray(node)) {
-            return node.length;
+            return node;
         }
 
-        return (node as JsonCompatibleObject)[key as LocalFieldKey] === undefined
-            ? 0 // A field with an undefined value has 0 length
-            : 1; // All other fields have a length of 1
-    }
-
-    public currentFieldLength(): number {
-        const parent = this.parentStack[this.parentStack.length - 1];
-        if (this.currentKey === EmptyKey && Array.isArray(parent)) {
-            return parent.length;
+        if (Object.prototype.hasOwnProperty.call(node, key)) {
+            const field = (node as JsonCompatibleObject)[key as LocalFieldKey];
+            return field === undefined ? [] : [field];
```

Comment:
Since the node object is supposed to be json compatible, and explicit undefined fields are not preserved in json (and make keysFromNode incorrect due to empty fields), I think we should assert not undefined here instead of handling it.